Add governance action deposits to stake-address-info query

CarlosLopezDeLara opened the pull request:
# Changelog

```yaml
- description: |
    Add governance action deposits to the output of `query stake-address-info` 
    
# uncomment types applicable to the change:
  type:
   - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
   - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Resolves https://github.com/IntersectMBO/cardano-cli/issues/954
Requires https://github.com/IntersectMBO/cardano-api/pull/730

# How to trust this PR

The query now includes a map of (GovActionIds, Deposit) assocoated to the stake address: 

```
❯ cardano-cli conway query stake-address-info --address stake_test1upumkxq6k4th9wtp4zp69gjwwwj5zmskxd6czlyf3xflzgq8yk95h
[
    {
        "address": "stake_test1upumkxq6k4th9wtp4zp69gjwwwj5zmskxd6czlyf3xflzgq8yk95h",
        "delegationDeposit": 2000000,
        "govActionDeposits": [
            {
                "1166b7591577807470bd527edf2a5fac938b8d56842a67684b682b1e8e76d198#0": 100000000000,
                "2f1be5c1aa824dac76b98acb35176defa6357e529cd297ae24353943b46a0cde#0": 100000000000,
                "d098afe0db98605c243c13c8a537a3eb51e6ded5e3a48acca83e7082a0086428#0": 100000000000
            }
        ],
        "rewardAccountBalance": 1194509145192,
        "stakeDelegation": "pool18pn6p9ef58u4ga3wagp44qhzm8f6zncl57g6qgh0pk3yytwz54h",
        "voteDelegation": "keyHash-74519ac5359c003a7ace69c475ae55a86eb8f9fec6cf6feaada1debf"
    }
]
```
or empty if it is not associated to any

```
❯ cardano-cli conway query stake-address-info --address $(< stake.addr)
[
    {
        "address": "stake_test1upfpm2244k8jf00l357t3adp2hzfsuqrwqvleheqjj08uhswme5cn",
        "delegationDeposit": 2000000,
        "govActionDeposits": [
            {}
        ],
        "rewardAccountBalance": 300000000000,
        "stakeDelegation": "pool1l9u9ss9xtww8qkt4zqda84z945f6tgq4753jqhtdr4r8yaw7d6g",
        "voteDelegation": "scriptHash-59aa3f091b3bcef254abfb89aea64973a61b78fdb2ac44839c7ccba8"
    }
]
```

**BABBAGE:** With the latest chantes the query still works in Babbage era: 

```
cardano-cli babbage query stake-address-info --address $(< example/utxo-keys/stake1.addr)
WARNING: Selected era is deprecated and will be removed in the future.
[
    {
        "address": "stake_test1uq07mnmf838hdlu9qpnwjnntrrtvsa6448ud0g32qkm0kecusq4sd",
        "delegation": "pool1xj26f4rqk2t049yu6n4mmg5zstr6sarv63nc5fuw60rx7u0jxef",
        "delegationDeposit": 2000000,
        "rewardAccountBalance": 0
    }
]
```


# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

<!--
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you.
-->

carbolymer committed:
Add certs to compatible transaction build command

* Added reference test for `compatible conway transaction singed-transaction`

CarlosLopezDeLara committed:
Add governance action deposits to stake-address-info query

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Query.hs:
```diff
@@ -970,17 +971,30 @@ callQueryStakeAddressInfoCmd
           easyRunQuery (queryStakeAddresses sbe stakeAddr networkId)
 
-        beo <- requireEon BabbageEra era
+        ceo <- requireEon ConwayEra era
```
So this will throw an error when we're in era < Conway. I think we should hide the command for eras < Conway in `pQueryStakeAddressInfoCmd` parser and add `ConwayEraOnwards era` field to `QueryStakeAddressInfoCmdArgs`.
(I'm not sure why it wasn't hidden for < Babbage in the first place 🤔)

I'm also wondering if it's not used in hardforking tests, may be worth double checking with @mkoura .

Jimbo4350 replied:
What we should do is start implementing and propagating queries that use the new api. Currently we have:

```
queryPoolState
  :: ()
  => BabbageEraOnwards era
  -> Maybe (Set PoolId)
 ....
```

Creating a separate module and parameterizing these queries on the new `data Era era` would be a good start. 

>  I'm also wondering if it's not used in hardforking tests, may be worth double checking with @mkoura .

Yes, we would need to confirm which commands need to retain backwards compatibility for the hardforking tests and any other tests that @mkoura relies on this for. Eventually we will shift those query commands under the `compatible`  name space in `cardano-cli`.

carbolymer replied:
Created: https://github.com/IntersectMBO/cardano-api/issues/732

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Query.hs:
```diff
@@ -970,17 +971,30 @@ callQueryStakeAddressInfoCmd
 
-        stakeVoteDelegatees <- monoidForEraInEonA era $ \ceo ->
-          easyRunQuery (queryStakeVoteDelegatees ceo stakeAddr)
+        govActionStates :: (Seq.Seq (L.GovActionState (ShelleyLedgerEra era))) <-
```
This field could be made optional and only show up in eras > Conway. This way the query would still be usable in Babbage.

CarlosLopezDeLara replied:
With the latest chantes the query still works in Babbage era:
```
cardano-cli babbage query stake-address-info --address $(< example/utxo-keys/stake1.addr)
WARNING: Selected era is deprecated and will be removed in the future.
[
    {
        "address": "stake_test1uq07mnmf838hdlu9qpnwjnntrrtvsa6448ud0g32qkm0kecusq4sd",
        "delegation": "pool1xj26f4rqk2t049yu6n4mmg5zstr6sarv63nc5fuw60rx7u0jxef",
        "delegationDeposit": 2000000,
        "rewardAccountBalance": 0
    }
]
```

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Query.hs:
```diff
@@ -1028,6 +1043,7 @@ writeStakeAddressInfo
                     , "voteDelegation" .= fmap friendlyDRep mDRep
                     , "rewardAccountBalance" .= mBalance
                     , "delegationDeposit" .= mDeposit
+                    , "govActionDeposits" .= [gaDeposits]
```
why a map is wrapped in a list?

carbolymer replied:
the `caseShelleyToBabbageOrConwayEraOnwards` now becomes useless, because babbage branch becomes dead code.

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Query.hs:
```diff
@@ -752,8 +752,8 @@ runQueryRefScriptSizeCmd
 newtype RefInputScriptSize = RefInputScriptSize {refInputScriptSize :: Int}
-  deriving Generic
-  deriving anyclass ToJSON
+  deriving (Generic)
```
We need to diagnose why this is happening. Something is up with your editor. Remove these non-semantic changes.

CarlosLopezDeLara replied:
this is being suggested when running `scripts/devshell/pretiffy `. I reverted it

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Query.hs:
```diff
@@ -937,6 +937,7 @@ runQueryStakeAddressInfoCmd
 data StakeAddressInfoData = StakeAddressInfoData
```
I think this type is now deserving of a haddock that outlines what the various records correspond to.

Jimbo4350 committed:
Introduce CliSpendScriptRequirements and subtypes with the aim of
removing all use of the ScriptWitnessFiles type

Jimbo4350 committed:
Remove usage of ScriptWitnessFiles WitCtxTxIn

Jimbo4350 committed:
Introduce and propagate pSimpleScriptOrPlutusSpendingScriptWitness and CliSpendScriptRequirements

Jimbo4350 committed:
pScriptWitnessFiles no longer handles spending script witnesses and
therefore we can remove the ShelleyBasedEra era parameter

Notice that as a result of factoring out the spending script witness
related code we have gained a simplification i.e we no longer have to
pollute several functions with the ShelleyBasedEra era parameter

Jimbo4350 committed:
Propagate SpendScriptWitness era within the transaction build/build-raw
commands

Jimbo4350 committed:
Misc changes

Jimbo4350 committed:
Implement Cardano.CLI.EraBased.Script.Types

Jimbo4350 committed:
Add certificate plutus script related types for use in parsing command
line arguments

CarlosLopezDeLara committed:
prettify

CarlosLopezDeLara committed:
temporary SRP

Jimbo4350 committed:
Replace use of readScriptWitnessFiles with readCertificateScriptWitnesses

Jimbo4350 committed:
Use CliCertificateScriptRequirements instead of ScriptWitnessFiles
WitCtxStake in compatibility transaction build command

Jimbo4350 committed:
Add to first commit! Creation of read module

Jimbo4350 committed:
Modify parsers for new certificate plutus scripts parsing types

CarlosLopezDeLara committed:
remove unnecessary parenthesis
